hazelcast/src/main/java/com/hazelcast/config/PermissionConfig.java

```diff
@@ -221,6 +229,62 @@ public PermissionConfig setEndpoints(Set<String> endpoints) {
         return this;
     }
 
+    @Override
+    public int getFactoryId() {
+        return ConfigDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return ConfigDataSerializerHook.PERMISSION_CONFIG;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeUTF(type.getNodeName());
+        out.writeUTF(name);
+        if (StringUtil.isNullOrEmptyAfterTrim(principal)) {
+            out.writeUTF("*");
+        } else {
+            out.writeUTF(principal);
+        }
+
+        out.writeInt(endpoints.size());
+        for (String endpoint : endpoints) {
+            out.writeUTF(endpoint);
+        }
+
+        out.writeInt(actions.size());
+        for (String action : actions) {
+            out.writeUTF(action);
+        }
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        type = PermissionType.getType(in.readUTF());
+        name = in.readUTF();
+        principal = in.readUTF();
+
+        int endpointsSize = in.readInt();
+        if (endpointsSize != 0) {
+            Set<String> endpoints = new HashSet<>();
```

Comment:
Minor: we can use `SetUtil.createHashSet()` here and below to avoid rehashing.

Reply:
Thanks for the suggestion. Changed that.

BTW `newSetFromMap(new ConcurrentHashMap<>())` initialization in `endpoints` and `actions` looks really weird to me, considering that these fields can be easily overwritten with a non-thread safe `Set` in their setters (and it's also done in `#readData`).

Reply:
It's weird indeed. We shouldn't need to deal with concurrent access in the config objects.

@kwart Do we still need the `newSetFromMap(...)`? My understanding is that it was a temporary fix until we merged https://github.com/hazelcast/hazelcast-enterprise/pull/2805. Can this be changed back to `HashSet`?